### PR TITLE
Enable smart nic ovs harward offloading

### PR DESCRIPTION
**Pull request by Bhavanaashok33**
Please refer : https://docs.google.com/document/d/16C_PBvxXY3d1-ZPQp_OCLWEARKK8u90QEpHJ9NyTuFM/edit#

- Using network-attachment-definition-client as third party resource which is maintain by k8snetworkplumbingwg
and their clientset and types.go are imported as a module.  https://github.com/k8snetworkplumbingwg/network-attachment-definition-client

- To obtain allocated resource information for  a pod, kubelet client queries pod resource service API and retrieve the pci address assigned to the pod 

User story:

1. Enable sriov_config in acc_provision input file which autogenerates sriov config map and daemonset and make its a part of aci_deployment file

sriov_config:
   enable: True         # default is False
   device_info:
         isRdma: False   # default is false
         devices: ""    # default is "1014","101e"

  2.Admin creates network attachment CRD with resource Name annotation referencing a resource pool from device plugin

apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata:
name: acicni
namespace: kube-system
annotations:
k8s.v1.cni.cncf.io/resourceName: mellanox.com/cx5_sriov_switchdev:
spec:
config: '{
"cniVersion": "0.3.1",
"name": "k8s-pod-network",
"plugins": [ { "type": "opflex-agent-cni","log_level": "debug", "ipam": { "type": "opflex-agent-cni-ipam" } }]
}

3. User supplies network CRD reference in the Pod annotation as well as resource requests from device resource pool 

apiVersion: v1
kind: Pod
metadata:
  name: offloadpod1
  labels:
    app: ovs-offload
  annotations:
     k8s.v1.cni.cncf.io/networks :  acicni
spec:
  containers:
   - name: ovs-offload-app
     image: networkstatic/iperf3
     imagePullPolicy: IfNotPresent
     command: ['sh', '-c', 'echo Container 1 is Running ; sleep 3600']
     resources:
       requests:
         mellanox.com/cx5_sriov_switchdev: '1'
       limits:
         mellanox.com/cx5_sriov_switchdev: '1'

4. Kubelet allocates a VF from requested SRIOV resource from the pool (With Topology manager support - Kubelet aligns VFs with Pod NUMA constraints)
5. ACI-CNI gets Pod Network attachment definition CRD from API-Server
6. ACI CNI queries pod resource service API and retrieve the pci address assigned to the pod 
7.  ACI CNI does the network plumbing where it moves the VF to pod's namespace
8. Upon pod deletion, ACICNI moves back the allocated VF of a pod to host's namespace

### CODE REVIEW

**Comment by Bhavanaashok33** [comment]
still working on this patch, please do not review yet

**Review by mchalla — APPROVED** [review]
Looks good to me.

**Comment by snaiksat** [comment]
Thanks @Bhavanaashok33 . The PR message had a fair bit of details, however the actual commit message which will go into the git repo's logs, does not have those. Could you please update the commit message to include these details. Also, note that the whoever reads the git logs will not have access to your document (it should not be referenced in the git commit message since its a private document), so mentioning the relevant details in the commit title and message is important.

**Comment by Bhavanaashok33** [comment]
@snaiksat Sure! Will do that

File: pkg/hostagent/setup.go [hunk 1]
```diff
@@ -431,10 +611,20 @@ func (agent *HostAgent) unconfigureContainerIfaces(id *md.ContainerId) error {
 	agent.deallocateMdIps(metadata)
 
 	logger.Debug("Clearing container interface")
+
 	for _, iface := range metadata.Ifaces {
-		err = runClearVeth(iface.Sandbox, iface.Name)
-		if err != nil {
-			logger.Error("Could not clear Veth ports: ", err)
+		if len(metadata.Id.DeviceId) != 0 {
+			logger.Debug("Moving the VF back to host's namespace")
```
Comment (tomflynn) [inline pkg/hostagent/setup.go:617]: Can you add the VF name to the log

File: pkg/hostagent/setup.go [hunk 2]
```diff
@@ -295,12 +463,24 @@ func (agent *HostAgent) configureContainerIfaces(metadata *md.ContainerMetadata)
...
+					iface.HostVethName, iface.Mac, iface.VfNetDevice, err =
+						runSetupVf(iface.Sandbox, iface.Name, mtu, ip.Address.IP, metadata.Id.DeviceId)
+					if err != nil {
+						fmt.Errorf("VF allocation failed :%v", err)
+					} else {
+						break
+					}
 				} else {
-					break
+					logger.Debug("Setting up veth")
```
Comment (tomflynn) [inline pkg/hostagent/setup.go:476]: Can you add veth name to the log?

File: pkg/hostagent/setup.go [hunk 3]
```diff
@@ -295,12 +463,24 @@ func (agent *HostAgent) configureContainerIfaces(metadata *md.ContainerMetadata)
 			//There are 4 cases: IPv4-only, IPv6-only, dual stack with either IPv4 or IPv6 as the first address.
 			//We are guaranteed to derive the MAC address from IPv4 if it is assigned
 			if ip.Address.IP != nil && ip.Address.IP.To4() != nil {
-				iface.HostVethName, iface.Mac, err =
-					runSetupVeth(iface.Sandbox, iface.Name, mtu, ip.Address.IP)
-				if err != nil {
-					return nil, err
+				if len(metadata.Id.DeviceId) > 0 && agent.config.OvsHardwareOffload {
+					logger.Debug("Setting up VF ")
```
Comment (tomflynn) [inline pkg/hostagent/setup.go:467]: Add VF details to log

File: pkg/hostagent/pods.go [hunk 4]
```diff
@@ -42,6 +42,7 @@ import (
 
 	uuid "github.com/google/uuid"
 	"github.com/noironetworks/aci-containers/pkg/metadata"
+	//md "github.com/noironetworks/aci-containers/pkg/metadata"
```
Comment (tomflynn) [inline pkg/hostagent/pods.go:45]: remove

**Review by tomflynn — APPROVED** [review]
I have a couple of minor log comments but it looks good to me

File: pkg/hostagent/setup.go (hunk 1 above)
Comment (Bhavanaashok33) [reply pkg/hostagent/setup.go:625]: addressed in the recent commit

File: pkg/hostagent/setup.go (hunk 2 above)
Comment (Bhavanaashok33) [reply pkg/hostagent/setup.go:481]: addressed in the recent commit

File: pkg/hostagent/setup.go (hunk 3 above)
Comment (Bhavanaashok33) [reply pkg/hostagent/setup.go]: addressed !

**Comment by Bhavanaashok33** [comment]
Patch is tried on overlay setup until today and is yet to be tested on ACI setup. Would like to merge once the testing is completed.

**Comment by Bhavanaashok33** [comment]
Closing this PR. There are many merge conflicts due to recent commits. I will create a new PR with the similar code changes with a new remote branch.

**Pull request closed without merging**